# Perkenalan

Nama: Bintang Muhammad <br>
Course: Data Science <br>
Batch: 6 <br>


# Judul/Penanda Soal

1. Central Tendency Data E - commerce
2. Hipothesis Testing di E - Commerce

# Import Library

In [260]:
#import libraries

import pandas as pd
import numpy as np
from scipy import stats
import datetime as dt

# Data Loading

In [261]:
#load data
df=pd.read_csv('https://raw.githubusercontent.com/fahmimnalfrzki/Dataset/main/purchase_data.txt?token=AEZDEHRU5YLQEXGDDLCL6CLA2KZTI', sep=';')

In [262]:
#preview data
df.head(5)

user_id                   timestamp      group landing_page  purchased
0   851104  2017-01-21 22:11:48.556739    control     old_page          0
1   804228  2017-01-12 08:01:45.159739    control     old_page          0
2   661590  2017-01-11 16:55:06.154213  treatment     new_page          0
3   853541  2017-01-08 18:28:03.143765  treatment     new_page          0
4   864975  2017-01-21 01:52:26.210827    control     old_page          1

Data Features Info

| Features | Description |
|-----------|-----------|
| user_id | id pelanggan
|timestamp | Waktu ketika pelanggan mengakses laman web
|group | grup kontrol dan treatment
|landing_page | landing page yang diakses oleh pelanggan (laman lama atau laman baru)
|purchased | 0 : tidak ada pembelian, and 1: ada pembelian

In [263]:
#view dataframe info
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 294478 entries, 0 to 294477
Data columns (total 5 columns):
 #   Column        Non-Null Count   Dtype 
---  ------        --------------   ----- 
 0   user_id       294478 non-null  int64 
 1   timestamp     294478 non-null  object
 2   group         294478 non-null  object
 3   landing_page  294478 non-null  object
 4   purchased     294478 non-null  int64 
dtypes: int64(2), object(3)
memory usage: 11.2+ MB


- Data terdiri dari 294478 baris
- Data memiliki 5 Fitur
- Tidak ada null values di data berdasarkan df info

In [264]:
#see uniqe values in group featoures
df['group'].unique()

array(['control', 'treatment'], dtype=object)

- This column only consist of 2 unique values

In [265]:
#see uniqe values in landing page featoures
df['landing_page'].unique()

array(['old_page', 'new_page'], dtype=object)

- This column only consist of 2 unique values

In [266]:
#see uniqe values in purchased featoures
df['purchased'].unique()

array([0, 1], dtype=int64)

- This column only consist of 2 unique values

# Data Processing

In [267]:
# check null values
df.isna().sum()

user_id         0
timestamp       0
group           0
landing_page    0
purchased       0
dtype: int64

- based on df.isna, there is no null values on the data

In [268]:
#check data type
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 294478 entries, 0 to 294477
Data columns (total 5 columns):
 #   Column        Non-Null Count   Dtype 
---  ------        --------------   ----- 
 0   user_id       294478 non-null  int64 
 1   timestamp     294478 non-null  object
 2   group         294478 non-null  object
 3   landing_page  294478 non-null  object
 4   purchased     294478 non-null  int64 
dtypes: int64(2), object(3)
memory usage: 11.2+ MB


- based on this info, there is 3 variables that must be changed to another data type, that is pruchased, timestamp and user id

In [269]:
#change data type to correct data type
df=df.astype({'purchased':'object','user_id':'object','timestamp':'datetime64[ms]'})

In [270]:
# data preview after data type changes
df.sample(5)

user_id               timestamp      group landing_page purchased
230932  687898 2017-01-03 00:58:40.956  treatment     new_page         0
138426  732393 2017-01-08 02:10:14.280  treatment     new_page         0
211107  727265 2017-01-16 14:58:25.965    control     old_page         0
46790   682293 2017-01-13 07:24:21.599    control     old_page         1
232965  935252 2017-01-15 22:18:57.098  treatment     new_page         0

In [271]:
#check data type again
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 294478 entries, 0 to 294477
Data columns (total 5 columns):
 #   Column        Non-Null Count   Dtype         
---  ------        --------------   -----         
 0   user_id       294478 non-null  object        
 1   timestamp     294478 non-null  datetime64[ns]
 2   group         294478 non-null  object        
 3   landing_page  294478 non-null  object        
 4   purchased     294478 non-null  object        
dtypes: datetime64[ns](1), object(4)
memory usage: 11.2+ MB


# Mathematical Calculation and Analysis

## Central Tendency

In [272]:
#view mode of purchased columns
print(df['purchased'].mode())
print(df['group'].mode())
print(df['landing_page'].mode())

0    0
dtype: object
0    treatment
dtype: object
0    new_page
1    old_page
dtype: object


- Most of the people did not buy on the e-commerce, regardless on the landing pages.
- Most of the user are a grouped treatement user
- Base on the mode for all visitor, tha data distribution is multimodal

In [273]:
#group buyer by date
df['date'] = df['timestamp'].dt.date
daily_buyer = df[['date','user_id']].groupby('date').count()

In [274]:
#view data
daily_buyer.head()

user_id
date               
2017-01-02     5783
2017-01-03    13394
2017-01-04    13284
2017-01-05    13124
2017-01-06    13528

In [275]:
#view meean and Median
print (daily_buyer['user_id'].mean())
print (daily_buyer['user_id'].median())

12803.391304347826
13381.0


- Based on the data, the average user who visits per day is 12803
- Based on the data, the median user who visits per day is 13381
- The mean is smaller than the median, so there is a possibility that the data is negatively skewed,even though it's not much, so there is also a possiblity that this is normally distributed data.  

In [276]:
#create new table to see how many people did not buy
df0=df[df['purchased']==0]

#view data
df0.head()

user_id               timestamp      group landing_page purchased  \
0  851104 2017-01-21 22:11:48.556    control     old_page         0   
1  804228 2017-01-12 08:01:45.159    control     old_page         0   
2  661590 2017-01-11 16:55:06.154  treatment     new_page         0   
3  853541 2017-01-08 18:28:03.143  treatment     new_page         0   
5  936923 2017-01-10 15:20:49.083    control     old_page         0   

         date  
0  2017-01-21  
1  2017-01-12  
2  2017-01-11  
3  2017-01-08  
5  2017-01-10

In [277]:
#group buyer by date
daily_buyer0 = df0[['date','user_id']].groupby('date').count()

In [278]:
#view data
daily_buyer0.head()

user_id
date               
2017-01-02     5073
2017-01-03    11872
2017-01-04    11701
2017-01-05    11565
2017-01-06    11902

In [279]:
#view meean and Median
print (daily_buyer0['user_id'].mean())
print (daily_buyer0['user_id'].median())

11271.347826086956
11800.0


- Based on the data, the average people who does not buy per day but visiting is 11271
- Based on the data, the median people who does not buy per day but visiting  is 11800
- The mean is smaller than the median, so there is a possibility that the data is negatively skewed, even though it's not much, so there is also a possiblity that this is normally distributed data.  

In [280]:
#create new table to see how many people did buy
df1=df[df['purchased']==1]

#view data
df1.head()

user_id               timestamp      group landing_page purchased  \
4   864975 2017-01-21 01:52:26.210    control     old_page         1   
6   679687 2017-01-19 03:26:46.940  treatment     new_page         1   
8   817355 2017-01-04 17:58:08.979  treatment     new_page         1   
9   839785 2017-01-15 18:11:06.610  treatment     new_page         1   
15  644214 2017-01-22 02:05:21.719    control     old_page         1   

          date  
4   2017-01-21  
6   2017-01-19  
8   2017-01-04  
9   2017-01-15  
15  2017-01-22

In [281]:
#group buyer by date
daily_buyer1 = df1[['date','user_id']].groupby('date').count()

In [282]:
#view dta
daily_buyer1.head()

user_id
date               
2017-01-02      710
2017-01-03     1522
2017-01-04     1583
2017-01-05     1559
2017-01-06     1626

In [283]:
#view meean and Median
print (daily_buyer1['user_id'].mean())
print (daily_buyer1['user_id'].median())

1532.0434782608695
1593.0


- Based on the data, the average buyer per day is 1532.04
- Based on the data, the median buyer is 1593
- The mean is smaller than the median, so there is a possibility that the data is negatively skewed, even though it's not much, so there is also a possiblity that this is normally distributed data.  

## A/B Testing

- Assumption: The original data is a sample data

In [284]:
#show the mean
print('average people who did buy:',daily_buyer1['user_id'].mean())
print('average people who did not buy:',daily_buyer0['user_id'].mean())

average people who did buy: 1532.0434782608695
average people who did not buy: 11271.347826086956


Critical Value: 0,05

Our hypothesis on this case:

**H0: μ_ people who buy = μ_ people who did not buy**

**H1: μ_people who buy != μ_people who did not buy**

In [285]:
#hypothesis testing
t_stat, p_val = stats.ttest_ind(daily_buyer1,daily_buyer0)

print('P-value:',p_val[0]) #the p-value isn't divided by 2 since the output is two-sided p-value
print('t-statistics:',t_stat[0])

P-value: 7.131181555098006e-29
t-statistics: -26.78047478845237


- The P Value is much higher than critical Value, so that means H0 is accepted
- That means the average of people who does not buy is does not significantly different tgan average of people who buy in the e-commerce

# Hasil

## Central Tendency

- Based on the mode Most of the people did not buy on the e-commerce(0 is dominant), regardless on the landing pages.
- Most of the user are a grouped treatement user
- Base on the mode, for all visitor tha data distribution is multimodal
- Based on the data, the average user who visits per day is 12803
- Based on the data, the median user who visits per day is 13381
- The mean is smaller than the median, so there is a possibility that the data is negatively skewed,even though it's not much, so there is also a possiblity that this is normally distributed data.  
- Based on the data, the average people who does not buy per day but visiting is 11271
- Based on the data, the median people who does not buy per day but visiting  is 11800
- The mean is smaller than the median, so there is a possibility that the data is negatively skewed, even though it's not much, so there is also a possiblity that this is normally distributed data. 
- Based on the data, the average buyer per day is 1532.04
- Based on the data, the median buyer is 1593
- The mean is smaller than the median, so there is a possibility that the data is negatively skewed, even though it's not much, so there is also a possiblity that this is normally distributed data. 

## AB Testing

- The P Value is much higher than critical Value, so that means H0 is accepted
- That means the average of people who does not buy is does not significantly different tgan average of people who buy in the e-commerce